check.go

```diff
@@ -0,0 +1,1086 @@
+					// The referenced layer should have a corresponding entry in
+					// one map or the other.
+					_, checked := referencedLayers[layer]
+					_, checkedRO := referencedROLayers[layer]
+					if !checked && !checkedRO {
+						err := fmt.Errorf("layer %s: %w", layer, ErrImageLayerMissing)
+						err = fmt.Errorf("%simage %s: %w", readWriteDesc, id, err)
+						if isReadWrite {
+							report.Images[id] = append(report.Images[id], err)
+						} else {
+							report.ROImages[id] = append(report.ROImages[id], err)
+						}
+					} else {
+						// Count this layer as referenced.  Whether by the
+						// image or one of its child layers doesn't matter
+						// at this point.
+						if _, ok := referencedLayers[layer]; ok {
+							referencedLayers[layer] = true
+						}
+						if _, ok := referencedROLayers[layer]; ok {
+							referencedROLayers[layer] = true
+						}
+					}
+					if isReadWrite {
+						if len(report.Layers[layer]) > 0 {
+							report.Images[id] = append(report.Images[id], report.Layers[layer]...)
+						}
+						if len(report.ROLayers[layer]) > 0 {
+							report.Images[id] = append(report.Images[id], report.ROLayers[layer]...)
+						}
+					} else {
+						if len(report.Layers[layer]) > 0 {
+							report.ROImages[id] = append(report.ROImages[id], report.Layers[layer]...)
+						}
+						if len(report.ROLayers[layer]) > 0 {
+							report.ROImages[id] = append(report.ROImages[id], report.ROLayers[layer]...)
+						}
+					}
+				}
+			}
+		}
+		return false, nil
+	}); errored {
+		return CheckReport{}, err
+	}
+
+	// Iterate over each container in turn.
+	if done, err := s.readContainerStore(func() (bool, error) {
+		containers, err := s.containerStore.Containers()
+		if err != nil {
+			return true, err
+		}
+		for i := range containers {
+			container := containers[i]
+			id := container.ID
+			logrus.Debugf("checking container %s", id)
+			if options.ContainerData {
+				// Check that all of the big data items are present and reading them
+				// back gives us the right amount of data.
+				for _, key := range container.BigDataNames {
+					func() {
+						data, err := s.containerStore.BigData(id, key)
+						if err != nil {
+							if errors.Is(err, os.ErrNotExist) {
+								err = fmt.Errorf("container %s: data item %q: %w", id, key, ErrContainerDataMissing)
+								report.Containers[id] = append(report.Containers[id], err)
+								return
+							}
+							err = fmt.Errorf("container %s: data item %q: %w", id, key, err)
+							report.Containers[id] = append(report.Containers[id], err)
+							return
+						}
+						if int64(len(data)) != container.BigDataSizes[key] {
+							err = fmt.Errorf("container %s: data item %q: %w", id, key, ErrContainerDataIncorrectSize)
+							report.Containers[id] = append(report.Containers[id], err)
+							return
+						}
+					}()
+				}
+			}
+			// Look at the container's base image.  If the image has errors, the image's errors
+			// are the container's errors.
```

Comment:
I’m not sure this is very useful — if we have a corrupt base layer, 20 application images build with that base layer, and 100 containers running those images, 100 individual reports of that layer being wrong seem rather redundant. _Especially_ if there were multiple reported errors per layer, or per image.

Maybe just a single “references an image/layer with errors” error would be sufficient.